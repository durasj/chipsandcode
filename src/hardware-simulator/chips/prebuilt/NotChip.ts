import type { ChipPin } from '../Chip';
import type Chip from '../Chip';
import IllegalStateError from '../IllegalStateError';

/**
 * NOT gate implemented natively
 */
class NotChip implements Chip {
  public readonly name = 'Not';

  private in: boolean = false;
  private out: boolean = true;

  setInput(name: string, value: boolean) {
    if (name !== 'in')
      throw new IllegalStateError(`Input pin '${name}' does not exist.`);

    this.in = value;
  }

  getOutput(name: string) {
    if (name !== 'out')
      throw new IllegalStateError(`Output pin '${name}' does not exist.`);
    return this.out;
  }

  getPins(): Map<string, ChipPin> {
    // Constructs map on the fly
    // Pin inspection is less performance sensitive than actual execution

    return new Map([
      ['in', { type: 'input', state: this.in, connections: [] }],
      ['out', { type: 'output', state: this.out, connections: [] }],
    ]);
  }

  public run() {
    this.out = !this.in;
  }
}

export default NotChip;
