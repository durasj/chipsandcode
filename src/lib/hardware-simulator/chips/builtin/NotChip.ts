import type { ChipPin } from '../Chip';
import type Chip from '../Chip';

/**
 * NOT gate implemented natively
 */
class NotChip implements Chip {
  public readonly name = 'Not' as string;

  private in: boolean[];
  private out: boolean[];

  constructor(private width = 1) {
    this.in = new Array(width).fill(false);
    this.out = new Array(width).fill(false);
  }

  setInput(_: string, value: boolean[]) {
    this.in = value;
  }

  getOutput() {
    return this.out;
  }

  getPins(): Map<string, ChipPin> {
    // Constructs map on the fly
    // Pin inspection is less performance sensitive than actual execution

    return new Map([
      ['in', { type: 'input', width: this.width, state: this.in, connections: [] }],
      ['out', { type: 'output', width: this.width, state: this.out, connections: [] }],
    ]);
  }

  public run() {
    this.out = this.in.map((b) => !b);
  }
}

export default NotChip;
