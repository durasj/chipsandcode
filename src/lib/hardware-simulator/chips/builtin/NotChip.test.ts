import { describe, expect, it } from 'vitest';

import NotChip from './NotChip';

describe('Not Chip', () => {
  it('Provides interface info', async () => {
    const chip = new NotChip();

    expect(chip.name).toBe('Not');
    expect(chip.getPins()).toMatchSnapshot();
  });

  it('Has correct logic', async () => {
    const chip = new NotChip();

    const values = [
      [[false], [true]],
      [[true], [false]],
    ];

    values.forEach(([input, out]) => {
      chip.setInput('in', input);
      chip.run();
      expect(chip.getOutput()).toStrictEqual(out);
    });
  });
});
