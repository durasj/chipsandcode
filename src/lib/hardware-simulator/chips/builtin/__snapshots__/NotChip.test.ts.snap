// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Not Chip > Provides interface info 1`] = `
Map {
  "in" => {
    "connections": [],
    "state": [
      false,
    ],
    "type": "input",
    "width": 1,
  },
  "out" => {
    "connections": [],
    "state": [
      false,
    ],
    "type": "output",
    "width": 1,
  },
}
`;
